Add FORMAT clause to convert datetime types to string and vice versa #2388

TreeHunter9 opened the pull request:
### 1. DATETIME TO STRING
The following flags are currently implemented for datetime to string conversion:
| Format Pattern | Description |
| ------------- | ------------- |
| YEAR | Year (1 - 9999) |
| YYYY | Last 4 digits of Year (0001 - 9999) |
| YYY | Last 3 digits of Year (000 - 999) |
| YY | Last 2 digits of Year (00 - 99) |
| Y | Last 1 digits of Year (0 - 9) |
| Q | Quarter of the Year (1 - 4) |
| MM | Month (01 - 12) |
| MON | Short Month name (Apr) |
| MONTH | Full Month name (APRIL) |
| RM | Roman representation of the Month (I - XII) |
| WW | Week of the Year (01 - 53) |
| W | Week of the Month (1 - 5) |
| D | Day of the Week (1 - 7) |
| DAY | Full name of the Day (MONDAY) |
| DD | Day of the Month (01 - 31) |
| DDD | Day of the Year (001 - 366) |
| DY | Short name of the Day (Mon) |
| J | Julian Day (number of days since January 1, 4712 BC) |
| HH / HH12 | Hour of the Day (01 - 12) with period (AM, PM)  |
| HH24 | Hour of the Day (00 - 23) |
| MI | Minutes (00 - 59) |
| SS | Seconds (00 - 59) |
| SSSSS | Seconds after midnight (0 - 86399) |
| FF1 - FF9 | Fractional seconds with the specified accuracy |
| TZH | Time zone in Hours  (-14 - 14) |
| TZM | Time zone in Minutes (00 - 59) |
| TZR | Time zone Name |

The dividers are:
| Dividers |
| ------------- |
| . |
| / |
| , |
| ; |
| : |
| 'space' |
| - |

Patterns can be used without any dividers:
```
SELECT CAST(CURRENT_TIMESTAMP AS VARCHAR(50) FORMAT 'YEARMMDD HH24MISS') FROM RDB$DATABASE;
=========================
20230719 161757
```
However, be careful with patterns like `DDDDD`, it will be interpreted as `DDD` + `DD`.  

It is possible to insert raw text into a format string with "": `... FORMAT '"Today is" DAY'` - Today is MONDAY.
Also the format is case-insensitive, so `YYYY-MM` == `yyyy-mm`.
Example:
```
SELECT CAST(CURRENT_TIMESTAMP AS VARCHAR(45) FORMAT 'DD.MM.YEAR HH24:MI:SS "is" J "Julian day"') FROM RDB$DATABASE;
=========================
14.6.2023 15:41:29 is 2460110 Julian day
```
### 2. STRING TO DATETIME
The following flags are currently implemented for string to datetime conversion:
| Format Pattern | Description |
| ------------- | ------------- |
| YEAR | Year |
| YYYY | Last 4 digits of Year |
| YYY | Last 3 digits of Year |
| YY | Last 2 digits of Year |
| Y | Last 1 digits of Year |
| MM | Month (1 - 12) |
| MON | Short Month name (Apr) |
| MONTH | Full Month name (APRIL) |
| RM | Roman representation of the Month (I - XII) |
| DD | Day of the Month (1 - 31) |
| J | Julian Day (number of days since January 1, 4712 BC) |
| HH / HH12 | Hour of the Day (1 - 12) with period (AM, PM)  |
| HH24 | Hour of the Day (0 - 23) |
| MI | Minutes (0 - 59) |
| SS | Seconds (0 - 59) |
| SSSSS | Seconds after midnight (0 - 86399) |
| FF1 - FF4 | Fractional seconds with the specified accuracy |
| TZH | Time zone in Hours  (-14 - 14) |
| TZM | Time zone in Minutes (0 - 59) |

Dividers are the same as for datetime to string conversion and can also be omitted.
Example:
```
SELECT CAST('2000.12.08 12:35:30.5000' AS TIMESTAMP FORMAT 'YEAR.MM.DD HH24:MI:SS.FF4') FROM RDB$DATABASE;
=====================
2000-12-08 12:35:30.5000
```

Commit:
Add FORMAT clause to convert datetime types to string and vice versa

Commit:
Add tests for FORMAT clause

sim1984 commented:
This functionality is in the SQL:2016 standard, so it would be nice to add it to Firebird. The only thing that seems to me it would be nice to expand it with named time zones.

omachtandras commented:
SELECT CAST(CURRENT_TIMESTAMP AS VARCHAR(45) FORMAT 'DD.MM.YEAR HH24:MI:SS "is" J "Julian day"') FROM RDB$DATABASE;
=========================
14.6.2023 15:41:29 is 2460110 Julian day

It would be good if it could be controlled in the case of the month and day so that it appears with a leading zero.

14.06.2023 15:41:29 is 2460110 Julian day

asfernandes commented:
> This functionality is in the SQL:2016 standard, so it would be nice to add it to Firebird. The only thing that seems to me it would be nice to expand it with named time zones.

Can you quote relevant SQL:2016 section?

Also, are you aware of other DBMS engine implementing this?

dyemanov commented:
Adriano: https://github.com/FirebirdSQL/firebird/issues/2388

AlexPeshkoff commented:
> It would be good if it could be controlled in the case of the month and day so that it appears with a leading zero.

+1
No matter that it's not covered by SQL standard.

aafemt commented:
AFAIR "MM" is used for zero-padded value and "M" for non-padded. The rest the same.

TreeHunter9 commented:
> AFAIR "MM" is used for zero-padded value and "M" for non-padded. The rest the same.

I thought about it, but it won't work with "DD", because "D" is already taken for Day of the Week. But maybe we can remove Day of the Week, if its not in standard, and use "D" for non-padded "DD".

sim1984 commented:
the problem concerns not only days and months. Minutes and seconds can also be with or without leading zero.

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -1033,6 +1047,841 @@ void CVT_string_to_datetime(const dsc* desc,
+			if (pattern == "TZH")
+			{
+				if (previousPattern == "TZM")
+					timezoneOffset *= sign(timezoneOffset);
```
Please explain this line.

TreeHunter9 replied:
I do this to remove the offset sign, so if we have the format "TZH:TZM", our output looks like "-03:30". But maybe I should use abs() for readability.

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -1033,6 +1047,841 @@ void CVT_string_to_datetime(const dsc* desc,
+	if (patternResult.isEmpty())
+		invalidPatternException(pattern, cb);
+
+	previousPattern = pattern;
```
Why are you assigning to a `string_view` on the function exit?

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -1033,6 +1047,841 @@ void CVT_string_to_datetime(const dsc* desc,
+	auto invalidPatternException = [](std::string_view pattern, Callbacks* cb)
+	{
+		string str(pattern.data(), pattern.length());
+		cb->err(Arg::Gds(isc_invalid_date_format) << Arg::Str(str.c_str()));
```
```suggestion
		cb->err(Arg::Gds(isc_invalid_date_format) << str);
```

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -1033,6 +1047,841 @@ void CVT_string_to_datetime(const dsc* desc,
+			break;
+
+		case dtype_sql_date:
+			Firebird::TimeStamp::decode_date(*(GDS_DATE *) desc->dsc_address, &times);
```
```suggestion
			Firebird::TimeStamp::decode_date(*(GDS_DATE*) desc->dsc_address, &times);
```

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -1033,6 +1047,841 @@ void CVT_string_to_datetime(const dsc* desc,
+		if (i < formatUpper.size() && separator != '\"')
+			result += separator;
+
+		offset = i + sizeof(separator);
```
```suggestion
		offset = i + 1;
```

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -1033,6 +1047,841 @@ void CVT_string_to_datetime(const dsc* desc,
+	{
+		int value = 0;
+
+		switch(str[offset])
```
```suggestion
		switch (str[offset])
```

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -1033,6 +1047,841 @@ void CVT_string_to_datetime(const dsc* desc,
+			{
+				if (previousPattern == "TZM")
+				{
+					outTimezoneInMinutes += sign(outTimezoneInMinutes)
+						* parse_string_to_get_int(str, strLength, strOffset, strLength - strOffset) * 60;
```
```suggestion
					outTimezoneInMinutes += sign(outTimezoneInMinutes) *
						parse_string_to_get_int(str, strLength, strOffset, strLength - strOffset) * 60;
```

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -1033,6 +1047,841 @@ void CVT_string_to_datetime(const dsc* desc,
+			{
+				if (previousPattern == "TZH")
+				{
+					outTimezoneInMinutes += sign(outTimezoneInMinutes)
+						* parse_string_to_get_int(str, strLength, strOffset, strLength - strOffset);
```
```suggestion
					outTimezoneInMinutes += sign(outTimezoneInMinutes) *
						parse_string_to_get_int(str, strLength, strOffset, strLength - strOffset);
```

dyemanov commented:
The standard does not seem to mention whether DD etc are zero-padded or not, but usually (in other languages) two-letter format abbreviation means exactly that.. So I'd suggest this being our default behaviour for all elements that are affected. This would also be consistent with our default date->char formatting:

```
select cast(cast('1.1.2020 1:1:1' as timestamp) as varchar(24)) from rdb$database;

CAST                     
======================== 
2020-01-01 01:01:01.0000 
```

note the leading zeroes.

dyemanov commented:
> This functionality is in the SQL:2016 standard, so it would be nice to add it to Firebird. The only thing that seems to me it would be nice to expand it with named time zones.

Oracle uses TZD for time zone abbreviations and TZR for time zone names. They're non-standard, but could be followed.

dyemanov commented:
I'm wondering where the non-standard format specifiers originate from? Oracle, I guess?

BTW, this is the default Oracle behaviour for padding:
 
> Oracle uses trailing blank characters and leading zeroes to fill format elements to a constant width. The width is equal to the display width of the largest element for the relevant format model:
> 
> Numeric elements are padded with leading zeros to the width of the maximum value allowed for the element. For example, the YYYY element is padded to four digits (the length of '9999'), HH24 to two digits (the length of '23'), and DDD to three digits (the length of '366').
> 
> The character elements MONTH, MON, DAY, and DY are padded with trailing blanks to the width of the longest full month name, the longest abbreviated month name, the longest full date name, or the longest abbreviated day name, respectively, among valid names determined by the values of NLS_DATE_LANGUAGE and NLS_CALENDAR parameters. For example, when NLS_DATE_LANGUAGE is AMERICAN and NLS_CALENDAR is GREGORIAN (the default), the largest element for MONTH is SEPTEMBER, so all values of the MONTH format element are padded to nine display characters. The values of the NLS_DATE_LANGUAGE and NLS_CALENDAR parameters are specified in the third argument to TO_CHAR and TO_* datetime functions or they are retrieved from the NLS environment of the current session.

So DD is 01, MM is 01, etc

Commit:
Fixes after review

TreeHunter9 commented on src/common/common.h:
```diff
@@ -993,6 +993,32 @@ static const TEXT* const FB_LONG_MONTHS_UPPER[] =
+};
+
+// Starts with SUNDAY cuz tm.tm_wday starts with it
+inline const TEXT* const FB_LONG_DAYS_UPPER[] =
```
There is no information about this in standard, but in Oracle, for example, Month and Day depends on language.

Commit:
Change TZD to TZR

asfernandes commented on src/common/common.h:
```diff
@@ -967,7 +967,7 @@ const int HIGH_WORD		= 0;
 #endif
 
-static const TEXT FB_SHORT_MONTHS[][4] =
+inline const TEXT FB_SHORT_MONTHS[][4] =
```
Look at `doc/README.modern_cpp.md`, we support a single feature of C++17. Only C++14 features can be generally used.

TreeHunter9 replied:
My bad, I'll change it back to static.

Commit:
Change inline variables back to static

asfernandes commented:
> I also have a question about truncating string, for example if we do `CAST(CURRENT_TIMESTAMP AS VARCHAR(20))` we get a truncation error (-expected length 20, actual 38). But what I do is that if the result > expected length, we truncate the result and assign what's left. Maybe I should throw the same truncation error?

I think so.

asfernandes commented:
> It is possible to insert raw text into a format string with "":

I do not have access to the SQL 2016 standard, and in https://github.com/FirebirdSQL/firebird/issues/2388 I see nothing about quotes.

Is this standard?

Isn't there a way to put literal double quotes in the output?

asfernandes commented:
This looks incorrectly in relation to time zone:
```
select
    current_timestamp a,
    cast(current_timestamp as varchar(50) format 'YEAR-MM-DD HH24:MI:SS.FF4') b,
    cast(cast(current_timestamp as varchar(50) format 'YEAR-MM-DD HH24:MI:SS.FF4') as timestamp with time zone format 'YEAR-MM-DD HH24:MI:SS.FF4') c
  from rdb$database;

A 2023-06-28 08:38:20.2790 America/Sao_Paulo
B 2023-06-28 08:38:20.2790
C 2023-06-28 08:38:20.2790 +00:00
```

In the case of `C`, it should use session time zone.

asfernandes commented:
Is it correct that tokens left in the trailing part is not matched with the format and do not generate error?

```
SELECT CAST('2000.12.08 12:35:30.5000 XXX' AS TIMESTAMP FORMAT 'YEAR.MM.DD HH24:MI:SS.FF4') FROM RDB$DATABASE;
```

asfernandes commented:
Please put the documentation (currently in the PR description) in a README formatted as markdown.

TreeHunter9 commented:
> Is this standard?

It's not in standard. I saw it in PostgreSQL and I thought it would be a good addition.
> Isn't there a way to put literal double quotes in the output?

No... I will fix it.

TreeHunter9 commented:
> Is it correct that tokens left in the trailing part is not matched with the format and do not generate error?
> 
> ```
> SELECT CAST('2000.12.08 12:35:30.5000 XXX' AS TIMESTAMP FORMAT 'YEAR.MM.DD HH24:MI:SS.FF4') FROM RDB$DATABASE;
> ```

From code perspective - it's correct, because we filled out our FORMAT, and there is no reason to continue parsing the input string.
But, perhaps, from the user's point of view there should be an error.

asfernandes commented on src/common/tests/CvtTest.cpp:
```diff
@@ -0,0 +1,398 @@
+	v.raise();
+}
+
+class CVTCallback : public Firebird::Callbacks
```
I don't think this is good way to share test code. You will need to declare things in others cpp files or use boost macros in header files.
If you need shared code, create cpp/h file for this code.

asfernandes commented on src/dsql/ExprNodes.cpp:
```diff
@@ -3465,8 +3473,16 @@ bool CastNode::setParameterType(DsqlCompilerScratch* /*dsqlScratch*/,
+	if (format.hasData())
+	{
+		dsqlScratch->appendUChar(blr_cast_format);
+		dsqlScratch->appendString(0, format);
```
You are ignoring the (implicit or explicit) connection charset (not `_WIN1252` in the example):
```
SELECT CAST(localTIMESTAMP AS VARCHAR(45) FORMAT _WIN1252 'DD.MM.YEAR HH24:MI:SS "is" J "Julian day"') FROM RDB$DATABASE;
```

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -1033,6 +1047,928 @@ void CVT_string_to_datetime(const dsc* desc,
+	if (format.isEmpty())
+		cb->err(Arg::Gds(isc_sysf_invalid_null_empty) << Arg::Str(STRINGIZE(format)));
+
+	auto invalidPatternException = [](std::string_view pattern, Callbacks* cb)
```
`std::string_view` is C++17 feature, not allowed to use in current Firebird code.
But since this should not go to Firebird 5, I will propose adoption of C++17 for Firebird 6 code.

TreeHunter9 replied:
I see other people agree to move to C++ 17, so I switch back static to inline variables.

Commit:
Add README documentation

Commit:
Add ability to use " in raw string and ...

Use session timezone if timezone is not specified.
Add ability to use + sign in timezone offset.
Add truncating string exception.

Commit:
Move util methods from BOOST_AUTO_TEST_SUITE

Commit:
Switch back to inline variables

TreeHunter9 committed:
Consider charset in the format string

TreeHunter9 committed:
Add ability to write patterns without separators

Commit:
Use printf to add extra zeros

Also add extra zeros to the year patterns.

aafemt commented:
I wonder why this code is so much complicated?

Template that is instantiated with only one type and only one callback parameter can be a plain function. Single-line callback for callback is mind-screwing.

to_upper() is locale-aware and is overkill in this case. If SQL standard doesn't require case insensitive format, I would make patterns case-sensitive.

Because you are not going to use locale-specific separators (do you?) everything that doesn't match pattern can be treated as plain text and appear in output as-is. In this case the parser is simplified to comparison with patterns starting from the longest and plain char* does the job.

TreeHunter9 commented:
> Template that is instantiated with only one type and only one callback parameter can be a plain function. Single-line callback for callback is mind-screwing.

Yea, it would be better.

> to_upper() is locale-aware and is overkill in this case. If SQL standard doesn't require case insensitive format, I would make patterns case-sensitive.

It's not mentioned in standard, but I think it gives more freedom when using format patterns.
If others think its unnecessary, I can make it case-sensitive.

Commit:
Replace template exception with a plain function

Commit:
Merge branch 'master' into master_cast_format_datetime_to_string_and_vice_versa

asfernandes commented on doc/README.cast.format.md:
```diff
@@ -0,0 +1,94 @@
+| SSSSS | Seconds after midnight (0 - 86399) |
+| FF1 - FF4 | Fractional seconds with the specified accuracy |
+| TZH | Time zone in Hours  (-14 - 14) |
+| TZM | Time zone in Minutes (0 - 59) |
```
Why `TZR` is not supported here?

TreeHunter9 replied:
I forgot about this... But this is very tricky to implement, because time zone names contains separators, for example Canada/East-Saskatchewan - it contains '/' and '-', and it's not easy to parse. I've written an implementation of this, but maybe it's too overwhelming for this kind of task. If you think it's not a good solution, I'll try to think of something else.

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -158,6 +177,12 @@ class DummyException {};
 static const double eps_float  = 1e-5;
 
+template <typename T>
+constexpr int sign(T value)
```
Please use `static`, enclose it in unnamed namespace or move it to a public generic header.

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -1033,6 +1058,964 @@ void CVT_string_to_datetime(const dsc* desc,
+			cb->err(Arg::Gds(isc_invalid_data_type_for_date_format));
+	}
+
+	string formatUpper(format);
```
```suggestion
	const string formatUpper(format);
```
As you save it in `string_view`.

asfernandes replied:
Sorry, I see you change it. Hope it's never reallocated.

TreeHunter9 replied:
It shouldn't

asfernandes commented on src/common/cvt.cpp:
```diff
@@ -1033,6 +1058,964 @@ void CVT_string_to_datetime(const dsc* desc,
+			}
+			else if (pattern == "SSSSS")
+			{
+				constexpr int maximumSecondsInDay = 23 * 60 * 60 + 59 * 60 + 59;
```
I don't understand that name/value. Max seconds in a day is `24 * 60 * 60`.
You write it with -1, maybe ok, but isn't clear to use `24 * 60 * 60 - 1` instead?

asfernandes commented on src/dsql/ExprNodes.cpp:
```diff
@@ -3481,24 +3481,30 @@ dsc* BoolAsValueNode::execute(thread_db* tdbb, Request* request) const
 	  dsqlField(aDsqlField),
 	  source(aSource),
 	  itemInfo(NULL),
+	  format(pool, format),
```
I don't personally agree, but we don't use parameters and fields with same name, so please do like the others and rename the parameter to `aFormat`.

asfernandes commented on src/dsql/ExprNodes.h:
```diff
@@ -242,7 +242,8 @@ class BoolAsValueNode final : public TypedNode<ValueExprNode, ExprNode::TYPE_BOO
 public:
-	explicit CastNode(MemoryPool& pool, ValueExprNode* aSource = NULL, dsql_fld* aDsqlField = NULL);
+	explicit CastNode(MemoryPool& pool, ValueExprNode* aSource = NULL, dsql_fld* aDsqlField = NULL,
+		const Firebird::string& format = NULL);
```
Rename here too.

asfernandes commented on src/dsql/parse.y:
```diff
@@ -695,6 +695,7 @@ using namespace Firebird;
 %token <metaNamePtr> TIMEZONE_NAME
 %token <metaNamePtr> UNICODE_CHAR
 %token <metaNamePtr> UNICODE_VAL
+%token <metaNamePtr> FORMAT
```
Please move it before `LOCKED`, i.e., alphabetical order.

asfernandes replied:
Sorry, instead, rebase it with master and put in section `// tokens added for Firebird 6.0`.

asfernandes replied:
Section is there already :), move it.

Commit:
Clean code after review

Commit:
Fix bug with TZH:TZM when TZH is 0

Commit:
Add TZR to STRING to DATE
